utils/include/utils/content_loader.h

```diff
@@ -35,15 +35,26 @@ class content_loader
     virtual std::optional<std::string> load_text(const resource_location& res_loc) const = 0;
     virtual list_directory_result list_directory_files(
         const utils::resource::resource_location& directory_loc) const = 0;
+    virtual list_directory_result list_directory_subdirs_and_symlinks(
+        const utils::resource::resource_location& directory_loc) const = 0;
     virtual std::string filename(const utils::resource::resource_location& res_loc) const = 0;
+    virtual bool file_exists(const utils::resource::resource_location& res_loc) const = 0;
+    virtual bool dir_exists(const utils::resource::resource_location& res_loc) const = 0;
+    virtual std::string lexically_relative(
+        const utils::resource::resource_location& p, const utils::resource::resource_location& q) const = 0;
 
 protected:
     ~content_loader() = default;
 };
 
 std::optional<std::string> load_text(const resource_location& res_loc);
 list_directory_result list_directory_files(const utils::resource::resource_location& directory_loc);
+list_directory_result list_directory_subdirs_and_symlinks(const utils::resource::resource_location& directory_loc);
 std::string filename(const utils::resource::resource_location& res_loc);
+bool file_exists(const utils::resource::resource_location& res_loc);
+bool dir_exists(const utils::resource::resource_location& res_loc);
+std::string lexically_relative(
```

Comment:
I'd like to discuss where to put `lexically_*(...)` type of things and I have two proposals:

1. I can rename `content_loader` to something like `resource_operations` and leave it there. The derived classes would be renamed as well. 
2. The `lexically_*(...)` functions could be stored in `resouce_location`. `resource_location` would then also contain some logic to re-route `lexically` requests to different handlers for e.g. filesystem behavior and specific URL behaviors.

How do you @slavek-kucera feel about it?

Reply:
I'd say that the `lexically_*` functions should be able to operate without any knowledge of the underlying fs.